File: src/bitstamp.r
print("Libraries loading time: ")
Packages <- c("dplyr", "readr", "anytime")
libraries_loading_time <-system.time(
  suppressMessages(lapply(Packages, library, character.only = TRUE)))
print(libraries_loading_time)

print("Data loading time: ")
data_loading_time <- system.time(suppressMessages(bitstamp <- read_csv("data/bitstamp.csv")))

print(data_loading_time)

print("Preprocessing time: ")
preprocessing_time <- system.time(bitstamp_dt <- bitstamp %>%
                    mutate(Datetime = anytime(Timestamp)) %>%
                    mutate(YearMonth = format(Datetime, "%Y-%m")))
print(preprocessing_time)

N = 10

time_s = system.time(for (i in 1:N){
  t <- bitstamp_dt %>%
    group_by(YearMonth) %>%
    summarise(AvgClose = mean(Close), MaxHigh = max(High))})
print("Processing time: ")
processing_time <- time_s / N
print(processing_time)

suppressMessages(log <- read_csv("logs/log.csv"))
DateTime <- Sys.time() %>% as.POSIXct
log <- log %>%
  bind_rows(list("DateTime" = DateTime, "Language" = "R", "Dataset" = "Bitstamp", "TaskNames" = "Library loading", "TaskTypes" = "Loading", "TaskExecTimes" = libraries_loading_time[['elapsed']])) %>%
  bind_rows(list("DateTime" = DateTime, "Language" = "R", "Dataset" = "Bitstamp", "TaskNames" = "File loading", "TaskTypes" = "Execution", "TaskExecTimes" = data_loading_time[['elapsed']])) %>%
  bind_rows(list("DateTime" = DateTime, "Language" = "R", "Dataset" = "Bitstamp", "TaskNames" = "Preprocessing", "TaskTypes" = "Execution", "TaskExecTimes" = preprocessing_time[['elapsed']])) %>%
  bind_rows(list("DateTime" = DateTime, "Language" = "R", "Dataset" = "Bitstamp", "TaskNames" = "Processing", "TaskTypes" = "Execution", "TaskExecTimes" = processing_time[['elapsed']]))
write_csv(log, "logs/log.csv")

File: src/bitstamp.jl
# using Pkg
# Pkg.add(["DataFrames", "DataFramesMeta", "CSV"])
# Pkg.add(PackageSpec(url="https://github.com/FNj/Hose.jl.git", rev="indexing"))

lib_load_time = @elapsed using DataFrames, CSV, DataFramesMeta, Hose, Statistics, StatsBase, Dates
# import Base.Meta.@dump
# const dir = @__DIR__

optionally_convert_to_df(df::DataFrame) = df
optionally_convert_to_df(df) = df |> DataFrame

file_load_compilation_time = @elapsed CSV.File("data/bitstamp.csv", missingstring="NA") |> optionally_convert_to_df
file_load_time = @elapsed const bitstamp = CSV.File("data/bitstamp.csv", missingstring="NA") |> optionally_convert_to_df


file_load_compilation_time -= file_load_time

# size(bitstamp, 1) / 227496
# names(hflights)
# head(hflights)
# describe(hflights)

# macro sample(df, n::Int)
#     esc(:($df[sample(1:nrow($df), $n), :]))
# end

# macro sample(df, q::Real)
#     esc(:($df[sample(1:nrow($df), round(Int, $q * nrow($df))), :]))
# end

function preprocess(df)
    @hose df |> @transform(Datetime = Dates.unix2datetime.(:Timestamp)) |> @transform(YearMonth = Dates.yearmonth.(:Datetime))
end

preprocess_compilation_time = @elapsed preprocess(bitstamp)
preprocess_exec_time = @elapsed const bitstamp_dt = preprocess(bitstamp)
preprocess_compilation_time -= preprocess_exec_time

# describe(bitstamp_dt)

# describe(bitstamp_dt)
# @time @hose bitstamp_dt |> @where(first.(:YearMonth) .< 2018)

function process(df)
    @hose df |>
    @by(:YearMonth, AvgClose = mean(:Close), MaxHigh = maximum(:High))
end

process_compile_time = @elapsed process(bitstamp_dt[1:5, :])
const N = 10
times = zeros(N)
# GC.enable(false)
# for i=1:N
#     GC.gc()
#     sleep(0.01)
#     times[i] = @elapsed process(hflights)
# end
# GC.enable(true)
#
# mean_time_no_GC = mean(times)

for i=1:N
    times[i] = @elapsed process(bitstamp_dt)
end

process_exec_time = mean(times)
process_compile_time -= process_exec_time
# process_strrev(df) = @hose df |> @transform(TailNum2=reverse.(:TailNum))
#
# string_reversal_compile_time = @elapsed process_strrev(hflights)
# string_reversal_time = @elapsed process_strrev(hflights)

println("""Stats:
- Libraries loading time: $lib_load_time s,
- File loading compilation time: $file_load_compilation_time s,
- File loading time: $file_load_time s,
- Preprocessing compilation time: $preprocess_compilation_time s,
- Preprocessing execution time: $preprocess_exec_time s,
- Group by compilation time: $process_compile_time s,
- Group by execution time: $process_exec_time s.""")

using Dates
task_names = vcat(["Library loading"], fill("File loading", 2), fill("Preprocessing", 2), fill("Processing", 2))
task_types = vcat(["Loading"], repeat(["Compilation", "Execution"], 3))
task_exec_times = [lib_load_time, file_load_compilation_time, file_load_time, preprocess_compilation_time, preprocess_exec_time, process_compile_time, process_exec_time]
stats_df = DataFrame(DateTime=fill(Dates.now(), length(task_names)), Language=fill("Julia", length(task_names)), Dataset=fill("Bitstamp", length(task_names)), TaskNames=task_names, TaskTypes=task_types, TaskExecTimes=task_exec_times)

try
    vcat(CSV.read("logs/log.csv") |> DataFrame, stats_df)
catch
    stats_df
end |> CSV.write("logs/log.csv")
